adapters/ucfunnel/ucfunnel.go

```diff
@@ -0,0 +1,136 @@
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+	"net/http"
+)
+
+type UcfunnelAdapter struct {
+	URI string
+}
+
+func NewUcfunnelBidder(endpoint string) *UcfunnelAdapter {
+	return &UcfunnelAdapter{
+		URI: endpoint}
+}
+
+func (a *UcfunnelAdapter) Name() string {
+	return "ucfunnel"
+}
+
+func (a *UcfunnelAdapter) SkipNoCookies() bool {
+	return false
+}
+
+func (a *UcfunnelAdapter) MakeBids(internalRequest *openrtb.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if response.StatusCode == http.StatusBadRequest {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
+		}}
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
+		}}
+	}
+
+	var errs []error
+	var bidResp openrtb.BidResponse
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{err}
+	}
+
+	var bidReq openrtb.BidRequest
+	if err := json.Unmarshal(externalRequest.Body, &bidReq); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(5)
+	for _, sb := range bidResp.SeatBid {
+		for i := range sb.Bid {
+			bidType := getBidType(bidReq, sb.Bid[i].ImpID)
+			if (bidType == openrtb_ext.BidTypeBanner) || (bidType == openrtb_ext.BidTypeVideo) {
+				b := &adapters.TypedBid{
+					Bid:     &sb.Bid[i],
+					BidType: bidType,
+				}
+				bidResponse.Bids = append(bidResponse.Bids, b)
+			}
+		}
+	}
+	return bidResponse, errs
+}
+
+func (a *UcfunnelAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	errs := make([]error, 0, len(request.Imp))
+
+	// If all the requests were malformed, don't bother making a server call with no impressions.
+	if len(request.Imp) == 0 {
+		return nil, errs
+	}
+
+	partnerId, partnerErr := getPartnerId(request)
+	if partnerErr != nil {
+		return nil, partnerErr
+	}
+
+	reqJSON, err := json.Marshal(request)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json")
+
+	uri := a.URI + "/" + partnerId + "/request"
+	return []*adapters.RequestData{{
+		Method:  "POST",
+		Uri:     uri,
+		Body:    reqJSON,
+		Headers: headers,
+	}}, errs
+}
+
+func getPartnerId(request *openrtb.BidRequest) (string, []error) {
+	var ext ExtBidderUcfunnel
+	err := json.Unmarshal(request.Imp[0].Ext, &ext)
+	if err != nil {
+		return "", []error{err}
+	}
+	if len(ext.Bidder.PartnerId) == 0 || len(ext.Bidder.AdUnitId) == 0 {
+		return "", []error{}
```

Comment:
We also seem to be returning an empty error list here. Do we want to add an error with a significative message? Just asking, this is not required.
```
 74   func (a *UcfunnelAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
 75       errs := make([]error, 0, len(request.Imp))
 76  
 77       // If all the requests were malformed, don't bother making a server call with no impressions.
 78       if len(request.Imp) == 0 {
    + 	      errs = append(errs, fmt.Errorf("No impression in the bid request\n"))
 79           return nil, errs
 80       }
adapters/ucfunnel/ucfunnel.go
```